backend/controller/timeline/timeline.go

```diff
@@ -0,0 +1,122 @@
+	"github.com/TBD54566975/ftl/backend/schema"
+	"github.com/TBD54566975/ftl/internal/log"
+	"github.com/TBD54566975/ftl/internal/model"
+)
+
+type Service struct {
+	dal dal.DAL
+}
+
+func New(ctx context.Context, conn libdal.Connection, encryption *encryption.Service) *Service {
+	return &Service{dal: *dal.New(conn, encryption)}
+}
+
+type Log struct {
+	DeploymentKey model.DeploymentKey
+	RequestKey    optional.Option[model.RequestKey]
+	Msg           *ftlv1.StreamDeploymentLogsRequest
+}
+
+func (s *Service) RecordLog(ctx context.Context, log *Log) error {
+	err := s.dal.InsertLogEvent(ctx, &dal.LogEvent{
+		RequestKey:    log.RequestKey,
+		DeploymentKey: log.DeploymentKey,
+		Time:          log.Msg.TimeStamp.AsTime(),
+		Level:         log.Msg.LogLevel,
+		Attributes:    log.Msg.Attributes,
+		Message:       log.Msg.Message,
+		Error:         optional.Ptr(log.Msg.Error),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to insert log event: %w", err)
+	}
+	return nil
+}
+
+type Call struct {
+	DeploymentKey    model.DeploymentKey
+	RequestKey       model.RequestKey
+	ParentRequestKey optional.Option[model.RequestKey]
+	StartTime        time.Time
+	DestVerb         *schema.Ref
+	Callers          []*schema.Ref
+	Request          *ftlv1.CallRequest
+	Response         optional.Option[*ftlv1.CallResponse]
+	CallError        optional.Option[error]
+}
+
+func (s *Service) RecordCall(ctx context.Context, call *Call) {
+	logger := log.FromContext(ctx)
+	var sourceVerb optional.Option[schema.Ref]
+	if len(call.Callers) > 0 {
+		sourceVerb = optional.Some(*call.Callers[0])
+	}
+
+	var errorStr optional.Option[string]
+	var stack optional.Option[string]
+	var responseBody []byte
+
+	if callError, ok := call.CallError.Get(); ok {
+		errorStr = optional.Some(callError.Error())
+	} else if response, ok := call.Response.Get(); ok {
+		responseBody = response.GetBody()
+		if callError := response.GetError(); callError != nil {
+			errorStr = optional.Some(callError.Message)
+			stack = optional.Ptr(callError.Stack)
+		}
+	}
+
+	err := s.dal.InsertCallEvent(ctx, &dal.CallEvent{
+		Time:             call.StartTime,
+		DeploymentKey:    call.DeploymentKey,
+		RequestKey:       optional.Some(call.RequestKey),
+		ParentRequestKey: call.ParentRequestKey,
+		Duration:         time.Since(call.StartTime),
+		SourceVerb:       sourceVerb,
+		DestVerb:         *call.DestVerb,
+		Request:          call.Request.GetBody(),
+		Response:         responseBody,
+		Error:            errorStr,
+		Stack:            stack,
+	})
+	if err != nil {
+		logger.Errorf(err, "failed to record call")
+	}
+}
+
+func (s *Service) InsertLogEvent(ctx context.Context, log *dal.LogEvent) error {
+	err := s.dal.InsertLogEvent(ctx, log)
+	if err != nil {
+		return fmt.Errorf("failed to insert log event: %w", err)
+	}
+	return nil
+}
+
+func (s *Service) QueryTimeline(ctx context.Context, limit int, filters ...dal.TimelineFilter) ([]dal.TimelineEvent, error) {
```

Comment:
Ideally I think we wouldn't be exposing `dal` types directly here, but in practice I think that would be difficult to achieve.

Reply:
Yeah I agree. I was torn here since it would be a bit of duplication. maybe as I add more to the timeline it will make sense to fully decouple. I'll keep an eye on it